BaseURL support

bbcrddave opened the pull request:
This PR adds BaseURL support.

It currently has two modes - DVB and "Basic". Basic mode selects BaseURLs in the order they are specified in the manifest. DVB mode is compliant with DVB A168 section 10.8.2. DVB is automatically selected if the DVB profile is listed at the top level of the MPD, otherwise basic mode is used.

In all cases, if a BaseURL has failed it will not be retried. Failures are carried across manifest updates (unless the available BaseURLs have changed), but are reset at the end of a session.

BaseURL elements are supported at all levels of the manifest as specified in ISO23009-1.  Implementation of this, as well as DVB mode, has made the entire solution more complicated than it might have been, but has led to the most flexible solution.

I've done a reasonable amount of testing but will not have covered all use cases and manifests and could have missed something, so I'd appreciate any testing people can do.

I am out of the office until Thursday of next week so will be unable to respond until then, but I welcome feedback on this large, complex PR.

LloydW93 commented on src/dash/models/DashManifestModel.js:
```diff
@@ -773,6 +787,56 @@ function DashManifestModel() {
+                    // it doesn't really make sense to have relative and
+                    // absolute URLs at the same level, or multiple
+                    // relative URLs at the same level, so assume we are
+                    // done from this level of the MPD
```
I think it could potentially make sense to have relative and absolute URLs at the same level - you could have the first URL as relative (i.e. continue loading requests from the same domain as the MPD), then fallback to mirrors on other domains. This is more thinking out loud than something I'd want to do myself though.

bbcrddave replied:
I suppose it might make sense, but I think/hope that it's probably a contrived example.

This comment, and the accompanying implementation, was based on informative guidelines in DVB A168 11.9.5 which, among other useful information, implies that mixing relative and absolute BaseURL elements at the same level would reduce clarity.

LloydW93 commented on src/streaming/rules/baseUrlResolution/DVBSelector.js:
```diff
@@ -0,0 +1,147 @@
+
+        const topPriorityFilter = function (baseUrl, idx, arr) {
+            return !idx || (
+                (arr[0].dvb_priority && baseUrl.dvb_priority) &&
```
should these be ```arr[idx]``` rather than ```arr[0]```? If so, should ```!idx``` be ```idx !== 0``` (as index 0 would be falsy)

LloydW93 replied:
Never mind, I get it now, was just a bit confusing!

bbcrddave replied:
Yes, I can see this could be confusing but it is definitely correct - I should probably have put a description in.

For everyone else: by the time this filter is applied, the list has already been sorted into descending priority order, so we know that the first entry in the list must have the highest priority. The first is automatically selected, then the remaining entries are then selected only if their priority matches the first, yielding a list of all entries with the highest priority.

LloydW93 commented on src/streaming/utils/BaseURLSelector.js:
```diff
@@ -0,0 +1,131 @@
+    function setup() {
+        serviceLocationBlacklistController = BlacklistController(context).create({
+            updateEventName:        Events.SERVICE_LOCATION_BLACKLIST_CHANGED,
+            loadFailedEventName:    Events.FRAGMENT_LOADING_COMPLETED
```
While this reads strange, that does make sense... it would be could if the download error that is sent out to the client is annotated with whether more options are available. Or exposing URL_RESOLUTION_FAILED would likely achieve give the client the information it would need to make a decision on whether to do something.

AkamaiDASH commented:
@bbcrddave  -Took time to look at all the changes and nothing popped out as being an issue.  The code is extremely clean and well thought out so as long as you feel comfortable with your testing I am excited to get this merged.  :+1:

bbcrddave commented:
@AkamaiDASH I'm happy with the testing I've done here, so probably the best thing is to get it merged. Then we can see if anyone else can break it :wink:

I'll do a rebase on Monday, unless you fancy a fun Friday task.

AkamaiDASH commented:
@bbcrddave Rebase on monday ;) most likely.

bbcrddave commented:
Rebased, and fixed a whitespace issue which caused git to complain.

Travis seems to have borked when unzipping the build tools (electron) on one of the runs. The other run passed. This is not a failure of our tests or our code.
